File: nic/e2etests/infra/endpoint.py
import re
from infra.ns import NS
from infra.app import AppFactory


def SetUpNs(name, ep):
    namespace_name = "_".join([name, "NET" if ep["remote"] else "HOST"])
    print ("Bringing up namespace for EP %s %s %s" % (name, ep["macaddr"], ep["ipaddrs"][0]))
    ns = NS(namespace_name)
    ns.BindInterface(name, None if ep["remote"] else ep.get("encap_vlan"))
    ns.SetMacAddress(ep["macaddr"])
    ns.SetIpAddress(ep["ipaddrs"][0], 24)
    return ns
           

def _get_id_from_string(string):
    patterns= [r'\d+']
    for p in patterns:
        match= re.findall(p, string)
        return int(match[0])

def _is_ep_remote(endpoint):
    return endpoint["spec"]["interface-type"] != "lif"

def _get_network_from_config(nw_name, full_cfg):
    for network in full_cfg["Networks"]["networks"]:
        if network["meta"]["name"] == nw_name:
            return network
           
class Endpoint:
    
    def __init__(self, name, data, full_cfg):
        self._name = name
        ep_cfg = data
        self._mac_addr = ep_cfg["spec"]["mac-address"]
        self._ip_address = ep_cfg["spec"]["ipv4-address"].split("/")[0]
        self._prefix_len = 24 
        self._lif_id = 0
        self._port = 0
        if not _is_ep_remote(ep_cfg):
            self._encap_vlan = ep_cfg["spec"]["useg-vlan"] 
            self._lif_id = _get_id_from_string(ep_cfg["spec"]["interface"])
        else:
            network = _get_network_from_config(ep_cfg["spec"]["network-name"], full_cfg)
            self._encap_vlan = network["spec"].get("vlan-id")
            self._port = _get_id_from_string(ep_cfg["spec"]["interface"])
        self._local = not _is_ep_remote(ep_cfg)


    def Init(self, id=None):
        app = AppFactory.Get(self._name, id=id)
        app.AttachInterface(self._name)
        app.SetMacAddress(self._name, self._mac_addr)
        if self._encap_vlan:
            app.AddVlan(self._name, self._encap_vlan)
            app.SetMacAddress(self._name, self._mac_addr, self._encap_vlan)
        if self._ip_address:
            app.SetIpAddress(self._name, self._ip_address, self._prefix_len, self._encap_vlan)
        self._app = app
    
    def Delete(self):
        print ("Cleaning up endpoint : %s" % self._name)
        self._app.Stop()
    
    def Run(self, cmd, timeout=None, background=False):
        return self._app.RunCommand(cmd, timeout, background)
        
    def GetIp(self):
        return self._ip_address

    def GetMac(self):
        return self._mac_addr
    
    def PrintEpInformation(self):
        print("\t Endpoint Type      : ", "Local" if self._local else "Remote")
        self._app.PrintAppInformation()
        
if __name__ == "__main__":
    pass
    

File: nic/e2etests/go/infra/endpoint.go
package infra

import (
	"fmt"
	"log"
	"regexp"
	"strconv"
	"strings"

	"github.com/pensando/sw/nic/e2etests/go/cfg"
	"github.com/pensando/sw/nic/e2etests/go/common"
)

//EpInterface EpInterface struct
type EpInterface struct {
	ID         int
	Name       string
	IPAddress  string
	PrefixLen  int
	MacAddress string
	EncapVlan  int
}

//Endpoint Endpoint struct
type Endpoint struct {
	Name      string
	Remote    bool
	Interface EpInterface
	AppEngine AppEngine
}

func getIDFromString(str string) (int, error) {
	re := regexp.MustCompile("[0-9]+")
	return strconv.Atoi((re.FindAllString(str, 1))[0])
}

func getNetworkFromConfig(nwName string, fullCfg *cfg.E2eCfg) *cfg.Network {
	for _, network := range fullCfg.NetworksInfo.Networks {
		if network.NetworkMeta.Name == nwName {
			return &network
		}
	}

	return nil
}

//Init Endpoint Init
func (ep *Endpoint) Init(container bool) {
	fmt.Println("Initialzing endpoint : ", ep.Name)
	if container {
		ep.AppEngine = NewContainer(ep.Name, "", "")
	} else {
		ep.AppEngine = NewNS(ep.Name)
		ep.AppEngine.BringUp()
	}
	ep.AppEngine.AttachInterface(ep.Interface.Name)
	ep.AppEngine.SetMacAddress(ep.Interface.Name, ep.Interface.MacAddress, 0)
	if ep.Interface.EncapVlan != 0 {
		ep.AppEngine.AddVlan(ep.Interface.Name, ep.Interface.EncapVlan)
		ep.AppEngine.SetMacAddress(ep.Interface.Name, ep.Interface.MacAddress, ep.Interface.EncapVlan)
	}
	if ep.Interface.IPAddress != "" {
		ep.AppEngine.SetIPAddress(ep.Interface.Name, ep.Interface.IPAddress, ep.Interface.PrefixLen,
			ep.Interface.EncapVlan)
	}
}

//Delete Endpoint Delete
func (ep *Endpoint) Delete() {
	fmt.Println("Cleaning up endpoint : ", ep.Name)
	ep.AppEngine.Stop()
}

//RunCommand Endpoint Run command
func (ep *Endpoint) RunCommand(cmd []string, timeout int, background bool) (common.CmdHandle, error) {
	return ep.AppEngine.RunCommand(cmd, 0, background)
}

//GetIP Endpoint Get IP
func (ep *Endpoint) GetIP() string {
	return ep.Interface.IPAddress
}

//GetIP Endpoint Get IP
func (ep *Endpoint) GetNetwork() string {
	return ep.Interface.IPAddress + "/" + strconv.Itoa(ep.Interface.PrefixLen)
}

//PrintInformation Print Endpoint Information
func (ep *Endpoint) PrintInformation() {
	var epType string
	if ep.Remote {
		epType = "Remote"
	} else {
		epType = "Local"
	}
	fmt.Println("\t Endpoint Type      : ", epType)
	ep.AppEngine.PrintAppInformation()
}

//NewEndpoint Instantiate new endpoint
func NewEndpoint(name string, data cfg.Endpoint, fullCfg *cfg.E2eCfg) *Endpoint {

	intfID, err := getIDFromString(data.EndpointSpec.Interface)
	network := getNetworkFromConfig(data.EndpointSpec.NetworkName,
		fullCfg)

	if network == nil {
		log.Fatalln("Network not found in config!", data.EndpointSpec.NetworkName)
	}

	remote := data.EndpointSpec.InterfaceType != "lif"

	if err != nil {
		log.Fatalln(err)
	}

	intf := EpInterface{
		ID:         intfID,
		Name:       name,
		MacAddress: data.EndpointSpec.MacAddresss,
		IPAddress:  strings.Split(data.EndpointSpec.Ipv4Address, "/")[0],
		PrefixLen:  24,
	}
	if remote {
		intf.EncapVlan = network.NetworkSpec.VlanID
	} else {
		intf.EncapVlan = data.EndpointSpec.UsegVlan
	}

	return &Endpoint{
		Name:      name,
		Remote:    remote,
		Interface: intf,
	}
}
